core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -467,6 +469,9 @@ class Partition(val topicPartition: TopicPartition,
       if (logManager.getLog(topicPartition, isFuture = true).isDefined)
         logManager.asyncDelete(topicPartition, isFuture = true)
     }
+    // If the deleted replica was leader replica, there might be delayed requests in purgatory that need to be cleaned up.
+    if (wasLeader)
+      tryCompleteDelayedRequests()
```

Comment:
I think we want to complete the delayed operations any time we receive the StopReplica, not just when we were a previous leader. With KIP-392, even followers may have fetches in purgatory that need to be completed. Also, I think this is useful even in the context of controlled shutdown, though usually we expect to see LeaderAndIsr requests received first.